### PR TITLE
RPC input validation Fixes #1611

### LINKED ISSUE #1611: rpc manager should validate RPC input types
using the new jsdoc parsing feature #1610 

- [x] validate and enforce types in RPC manager. 
- [x] parse the inputs to correct types
- [x] provide feedback to the user

### PR DESCRIPTION
**Pull request by hamidzr**
- validates input types
- casts inputs to their correct type before calling the associated function
- provides feedback to the user
- checks for whether the argument is optional or not

### CODE REVIEW

File: src/client/blocks-ext.js [hunk 1]
```diff
@@ -256,9 +256,20 @@ RPCInputSlotMorph.prototype.methodSignature = function () {
     if (rpc) {
         this.fieldsFor = JSON.parse(this.getURL('/rpc/' + rpc));
 
+        var privateFields = {};
+
+        // hide/extract private attributes
+        for (let key in this.fieldsFor) {
```
Comment (brollb) [inline src/client/blocks-ext.js]: don't use `let` on the client

File: src/client/blocks-ext.js [hunk 2]
```diff
@@ -256,9 +256,20 @@ RPCInputSlotMorph.prototype.methodSignature = function () {
     if (rpc) {
         this.fieldsFor = JSON.parse(this.getURL('/rpc/' + rpc));
 
+        var privateFields = {};
```
Comment (brollb) [inline src/client/blocks-ext.js]: I am not sure if I like this approach for serving the service info. Rather than adding fields prefixed with a `_` to encode additional information for the entire service, it might be better to change the overall signature in a more easily parsed and intuitive way such as returning the following for the `'/rpc/' + rpc` endpoint:

```
[
  {
    "name": "getTerrainMap",
    "arguments": ["latitude", "longitude", "zoom", "width", "height"],
    "optional": [],
    "deprecated": false
  }
]
```

File: src/server/rpc/procedures/chart/chart.js [hunk 3]
```diff
@@ -96,6 +96,12 @@ function genGnuData(lines, lineTitles, lineTypes, smoothing){
     });
 }
 
+/**
+ * draw charts and histograms
+ * @param {Array} lines a single line or list of lines. Each line should be in form of [[x1,y1], [x2,y2]]
+ * @param {Array=} options optional options
```
Comment (brollb) [inline src/server/rpc/procedures/chart/chart.js]: The description here doesn't provide a whole lot of useful information. I would consider saying something like "Configuration for graph title, axes, and more" or something

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 4]
```diff
@@ -135,7 +140,14 @@ if(!process.env.GOOGLE_GEOCODING_API) {
             });
     };
 
-    // find places near a coordinate (20 reults max)
+    /**
+     * find places near a coordinate (20 reults max)
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: typo: `reults`

I would also consider being more clear about things like "coordinate" because that doesn't have a lot of meaning in the current form (especially when the target audience is young).

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 5]
```diff
@@ -135,7 +140,14 @@ if(!process.env.GOOGLE_GEOCODING_API) {
...
+     * @param {Latitude} latitude 
+     * @param {Longitude} longitude
+     * @param keyword
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: This is missing a type

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 6]
```diff
@@ -135,7 +140,14 @@ if(!process.env.GOOGLE_GEOCODING_API) {
...
+     * @param {Number} radius search radius in meters
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: Isn't this optional?

File: src/server/rpc/procedures/thing-speak/thing-speak.js [hunk 7]
```diff
@@ -137,7 +137,12 @@ thingspeakIoT.privateChannelFeed = function(id, numResult, apiKey) {
     }
 };
 
-//put together the data from feeds and channel metadata
+/**
+ * put together the data from feeds and channel metadata
```
Comment (brollb) [inline src/server/rpc/procedures/thing-speak/thing-speak.js]: I would reword this description to be more informative, concise and fit more with the tone/style of the other descriptions.

File: src/server/rpc/procedures/thing-speak/thing-speak.js [hunk 8]
```diff
@@ -137,7 +137,12 @@ thingspeakIoT.privateChannelFeed = function(id, numResult, apiKey) {
...
+ * @param {Number} id channel ID
+ * @returns {Object}
```
Comment (brollb) [inline src/server/rpc/procedures/thing-speak/thing-speak.js]: A description here would be nice

File: src/server/rpc/rpc-manager.js [hunk 9]
```diff
@@ -14,9 +14,55 @@ var fs = require('fs'),
...
+    let errorMsg;
+    const genericErr = `${arg.name} = ${input} is not a valid ${arg.type}. `;
+    if (input === '') {
+        if (!arg.isOptional) {
+            errorMsg = `${arg.name} is required.`;
+        }
+        input = undefined;
+        return [input, errorMsg];
+    }
+    // TODO add check for optional
```
Comment (brollb) [inline src/server/rpc/rpc-manager.js]: Don't you already have a check for optional arguments?

File: src/server/rpc/rpc-manager.js [hunk 10]
```diff
@@ -14,9 +14,55 @@ var fs = require('fs'),
...
+    switch (arg.type) {
```
Comment (brollb) [inline src/server/rpc/rpc-manager.js]: This would be better to move into a seperate file which returns an object of all the types where the keys of the object are the types and contains the casting functionality. This would allow the error message to be generated automatically rather than adding them everywhere.

File: src/server/rpc/rpc-manager.js (hunk 9 above)
Comment (hamidzr) [reply src/server/rpc/rpc-manager.js]: yep, it's leftover

File: src/server/rpc/rpc-manager.js (hunk 10 above)
Comment (hamidzr) [reply src/server/rpc/rpc-manager.js]: 8 lines above this I have a line that generates the messages based on the arg info automatically, and then I'm appending specific messages to each type, like the range for latitude and long. However, it would be good to move this to a separate module.

File: src/client/blocks-ext.js (hunk 2 above)
Comment (hamidzr) [reply src/client/blocks-ext.js]: should we have that in another PR?

Comment (brollb) [reply src/client/blocks-ext.js]: Yeah, that is probably a good idea.

File: src/server/rpc/rpc-manager.js [hunk 11]
```diff
@@ -215,6 +243,32 @@ RPCManager.prototype.handleRPCRequest = function(RPC, req, res) {
...
+                        args[idx] = input.value;
+                    } else {
+                        // handle the error
+                        this._logger.warn(`${rpc.serviceName} -> ${action} input error:`, input.msg);
+                        if (input.msg) errors.push(input.msg);
+                    }
+                }
+            });
+            // provide user feedback if there was an error
+            if (errors.length > 0) return res.status(500).send(errors.join(' '));
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply src/server/rpc/rpc-manager.js]: true, using \n makes for a better experience

**Comment by hamidzr** [comment]
I addressed the comments. I think it would be a good idea to programmatically capitalize sentences.

File: src/client/blocks-ext.js (hunk 2 above)
Comment (hamidzr) [reply src/client/blocks-ext.js]: I created an issue for this https://github.com/NetsBlox/NetsBlox/issues/1620

**Comment by brollb** [comment]
It's looking pretty good. There are two things that should probably be addressed:
- types should probably be converted to NetsBlox types in the error message
- there appears to be an exception thrown in the rpc-manager/type checking somewhere
deepinscreenshot_select-area_20171027173520

If it's not to hard to recreate this issue in a test, it also would be good.

**Comment by brollb** [comment]
Overall, it looks like it is almost good to go! However, there is one little issue with the tests. Tests should try to have one assert per tests (not a hard rule) and should test one very precise thing. Ideally, this is reflected in the name of the test. I updated one of the tests as an example.

Does that make sense?

**Comment by hamidzr** [comment]
yes, not doing it like that was just a cost-benefit analysis of a more descriptive test description.

**Comment by brollb** [comment]
Looks good now! Nice work.
